NIFI-11958 Azure Data Explorer as sink

tanmaya-panda1 opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-11958](https://issues.apache.org/jira/browse/NIFI-11958)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [ ] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [ ] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [ ] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

tanmaya-panda1 commented:
@exceptionfactory : Can you please take a look at this PR for Azure Data Explorer acting as sink.

joewitt commented:
@tanmaya-panda1 Hello.  Can you please rename the processor to be more specific/clear to what it is doing.  If we're sending to ADX a name such as 'PublishAzureDataExplorer' is more clear in intent as "IngestAzureDataExplorer" could mean we're getting data from it OR we're sending data to it.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/IngestAzureDataExplorer.java:
```diff
@@ -0,0 +1,401 @@
+
+            KustoIngestionResult result = service.ingestData(new KustoIngestionRequest(isStreamingEnabled, pollOnIngestionStatus, inputStream, ingestionProperties));
+            getLogger().info("Poll on Ingestion Status {} and Operation status: {} ", pollOnIngestionStatus, result.toString());
+            if (result == KustoIngestionResult.SUCCEEDED) {
+                getLogger().info("Operation status Succeeded - {}", result.toString());
+            }
+
+            if (result == KustoIngestionResult.FAILED) {
+                getLogger().error("Operation status Error - {}", result.toString());
+                isError = true;
+            }
+
+            if (result == KustoIngestionResult.PARTIALLY_SUCCEEDED) {
+                getLogger().error("Operation status Partially succeeded - {}", result.toString());
+                isError = true;
+            }
```
These conditional should be linked together with `else if` instead of individual conditions.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-services-api/pom.xml:
```diff
@@ -36,6 +36,16 @@
             <groupId>org.apache.nifi</groupId>
             <artifactId>nifi-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure.kusto</groupId>
+            <artifactId>kusto-data</artifactId>
+            <version>5.0.1</version>
+        </dependency>
+        <dependency>
+            <groupId>com.microsoft.azure.kusto</groupId>
+            <artifactId>kusto-ingest</artifactId>
+            <version>5.0.1</version>
+        </dependency>
```
Are these additions necessary at the API package level? It would be better to structure the interface to avoid the need for these dependencies if possible.

tanmaya-panda1 replied:
yeah, I struggled to decouple the dependencies of azure-kusto-java sdk from API package level. The interfaces KustoIngestService and KustoQueryService are decoupled from kusto java SDK dependency. 
But had to include dependencies in the API package level since we had com.microsoft.azure.kusto.ingest.IngestionProperties while submitting an ingestion request to Kusto cluster and com.microsoft.azure.kusto.data.KustoResultSetTable as a part ingestionResult from kusto after ingestion.

exceptionfactory reviewed (changes requested):
Thanks for working on this new component @tanmaya-panda1. I noted some initial recommendations, but there are some important details to consider in terms of the implementation that will take more time to review.

tanmaya-panda1 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,314 @@
+        StreamSourceInfo info = new StreamSourceInfo(kustoIngestionRequest.getInputStream());
+        //ingest data
+        IngestionResult ingestionResult;
+        ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
```
Agreed, modified the ingestionStatusCheck to execute in the same thread

tanmaya-panda1 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoQueryService.java:
```diff
@@ -159,10 +162,7 @@ private ConnectionStringBuilder getConnectionStringBuilder(final ConfigurationCo
-        final String version = System.getProperty("java.version");
-
-        builder.setConnectorDetails(vendor, version, null, null, false, null);
+        builder.setConnectorDetails("Kusto.Nifi.Source", StandardKustoQueryService.class.getPackage().getImplementationVersion(), null, null, false, null, NIFI_SOURCE);
```
Yes this change is necessary since connector details/identifier along with version information is required for identification of request in ADX(Azure Data Explorer) cluster for debugging and tracking purpose. This is a standard across all the connectors we have implemented for Azure Data Explorer.

tanmaya-panda1 commented:
@exceptionfactory , Hope you are doing well. Can you please take a look the PR again.

exceptionfactory reviewed (changes requested):
Thanks for making the updates @tanmaya-panda1, that helped to clarify a few things. Thanks for your patience on this review, as there are a number of efforts in progress for NiFi 2.0 right now. I went through the latest version and provided a number of comments and recommendations.

At a high level, it still appears to be possible to decouple the `KustoIngestService` from the Kusto API with a few changes. Please see the detailed comments for recommendations.

If we can work through this latest round of feedback, that should cover the majority of remaining issues, and then we can focus on wrapping up any remaining details.

tanmaya-panda1 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,399 @@
+                getLogger().error("Operation status Error - {}", result.toString());
+                isError = true;
+            } else if (result == KustoIngestionResult.PARTIALLY_SUCCEEDED) {
+                getLogger().error("Operation status Partially succeeded - {}", result.toString());
```
Considering partially succeeded status as success might result in silent data loss when the user ignores warning logs for a huge set of data, Hence added a new "partial success" relationship where the partially succeeded data can be routed and appropriate action can be taken by the user. What are your thoughts on this?

exceptionfactory replied:
Thanks for flagging this scenario @tanmaya-panda1. What would the user be expected to do in this scenario? The problem seems to be knowing what was ingested and what was not ingested. One option might be to introduce a configurable property for `Partial Success Routing`, with values of `success` or `failure`. This would let the user decide on a high-level strategy. In addition, the status should be noted in a FlowFile attribute, which would support routing decisions that could include retrying.

tanmaya-panda1 commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-services-api/src/main/java/org/apache/nifi/services/azure/data/explorer/KustoIngestService.java:
```diff
@@ -0,0 +1,28 @@
+
+@Tags({"azure", "adx", "explorer", "kusto", "ingest"})
+@CapabilityDescription("Connection-Service for Azure ADX (Kusto) ingestion cluster to ingest data.")
+public interface KustoIngestService extends KustoQueryService {
```
While using KustoIngestService, we will need to query for ingestor privileges and then start ingestion hence we need executeQuery during the ingest service. We didn't want to duplicate executeQuery in KustoIngestService, so extended KustoQueryService.

exceptionfactory replied:
On further review of the response, it seems like it would be better to introduce a new method for executing the ingest status query, instead of returning the generic `List<List<Object>>` result.

tanmaya-panda1 commented:
@exceptionfactory : Thanks for the detailed review. I have addressed most of your comments, please feel free to review and comment if anything is missing. Thanks once again.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
```
The `InputRequirement` annotation should be added with `Requirement.INPUT_REQUIRED` specified.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+    public static final String STREAMING_POLICY_SHOW_COMMAND = ".show %s %s policy streamingingestion";
+    public static final String DATABASE = "database";
+
+    private List<PropertyDescriptor> descriptors;
+    private Set<Relationship> relationships;
+    private transient KustoIngestService service;
+    private boolean streamingEnabled;
+    private boolean pollOnIngestionStatus;
```
These member variables should be moved down, following after `static` values, to maintain common conventions.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            .build();
+
+    public static final PropertyDescriptor IS_STREAMING_ENABLED = new PropertyDescriptor
+            .Builder().name("Streaming enabled")
+            .displayName("Streaming enabled")
```
```suggestion
            .Builder().name("Streaming Enabled")
            .displayName("Streaming Enabled")
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,399 @@
+            .build();
+
+    public static final PropertyDescriptor POLL_ON_INGESTION_STATUS = new PropertyDescriptor
+            .Builder().name("Poll on ingestion status")
+            .displayName("Whether to Poll on ingestion status")
```
The name should be changed to use Title Case.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            .displayName("Route partially successful ingestion records")
+            .description("Defines where to route partially successful ingestion records.")
+            .required(false)
+            .allowableValues("Success", "Failure")
+            .defaultValue("Failure")
```
These values should reuse the available relationship names of `success` and `failure`.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            .build();
+
+    static final PropertyDescriptor INGESTION_STATUS_POLLING_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("Timeout for polling on ingestion status")
+            .displayName("Timeout for polling on ingestion status in seconds")
```
The name should be shorter and use Title Case.
```suggestion
            .name("Ingest Status Polling Timeout")
            .displayName("Ingest Status Polling Timeout")
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+    static final PropertyDescriptor INGESTION_STATUS_POLLING_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("Timeout for polling on ingestion status")
+            .displayName("Timeout for polling on ingestion status in seconds")
+            .description("Defines the value of timeout for polling on ingestion status in seconds")
+            .required(false)
+            .dependsOn(POLL_ON_INGESTION_STATUS, "true")
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("600")
```
The value should not be in seconds, but should use the available `TIME_PERIOD_VALIDATOR`, which supports expressions such as `5 minutes` or `30 s`.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            .build();
+
+    static final PropertyDescriptor INGESTION_STATUS_POLLING_INTERVAL = new PropertyDescriptor.Builder()
+            .name("Ingestion status polling interval")
+            .displayName("Ingestion status polling interval in seconds")
```
```suggestion
            .name("Ingest Status Polling Interval")
            .displayName("Ingest Status Polling Interval")
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            .description("Defines the value of timeout for polling on ingestion status in seconds.")
+            .required(false)
+            .dependsOn(POLL_ON_INGESTION_STATUS, "true")
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("5")
```
See above for recommendation on using the Time Period validator.
```suggestion
            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
            .defaultValue("5 s")
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            .displayName("Ingestion status polling interval in seconds")
+            .description("Defines the value of timeout for polling on ingestion status in seconds.")
+            .required(false)
+            .dependsOn(POLL_ON_INGESTION_STATUS, "true")
```
```suggestion
            .dependsOn(POLL_ON_INGESTION_STATUS, Boolean.TRUE.toString())
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        final List<PropertyDescriptor> descriptorList = new ArrayList<>();
+        descriptorList.add(ADX_SERVICE);
+        descriptorList.add(DATABASE_NAME);
+        descriptorList.add(TABLE_NAME);
+        descriptorList.add(MAPPING_NAME);
+        descriptorList.add(DATA_FORMAT);
+        descriptorList.add(IGNORE_FIRST_RECORD);
+        descriptorList.add(IS_STREAMING_ENABLED);
+        descriptorList.add(POLL_ON_INGESTION_STATUS);
+        descriptorList.add(ROUTE_PARTIALLY_SUCCESSFUL_INGESTION);
+        descriptorList.add(INGESTION_STATUS_POLLING_TIMEOUT);
+        descriptorList.add(INGESTION_STATUS_POLLING_INTERVAL);
+        this.descriptors = Collections.unmodifiableList(descriptorList);
+
+        final Set<Relationship> relationshipSet = new HashSet<>();
+        relationshipSet.add(SUCCESS);
+        relationshipSet.add(FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationshipSet);
```
These collections should be created in a static initializer instead, so this `init()` method can be removed.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+    public void onScheduled(final ProcessContext context) {
+        service = context.getProperty(ADX_SERVICE).asControllerService(KustoIngestService.class);
+        if (checkIfIngestorRoleDoesntExist(context.getProperty(DATABASE_NAME).getValue(), context.getProperty(TABLE_NAME).getValue())) {
+            getLogger().error("User might not have ingestor privileges, table validation will be skipped for all table mappings.");
+            throw new ProcessException("User might not have ingestor privileges, table validation will be skipped for all table mappings. ");
```
It is not necessary to both log an error and throw an exception. Recommend removing the error log and include the database and table name in the exception message.

The message says that validation will be skipped, but throwing an exception prevents the Processor from starting, so the behavior should be clarified.

tanmaya-panda1 replied:
Since without ingestor privilege, there is no point in starting the ingestion of data, hence we throw an exception. Also removed skipping of table validation message which might confuse the user.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+        streamingEnabled = context.getProperty(IS_STREAMING_ENABLED).evaluateAttributeExpressions().asBoolean();
+        if (streamingEnabled && !checkIfStreamingPolicyIsEnabledInADX(context.getProperty(DATABASE_NAME).getValue(), context.getProperty(DATABASE_NAME).getValue())) {
+            getLogger().error("Streaming policy is not enabled in database {}", context.getProperty(DATABASE_NAME).getValue());
+            throw new ProcessException("Streaming policy is not enabled in database {}" + context.getProperty(DATABASE_NAME).getValue());
```
The `{}` placeholder is for log messages, but that doesn't work for exception messages.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+
+        // ingestion into ADX
+        try (final InputStream inputStream = session.read(flowFile)) {
+            StringBuilder ingestLogString = new StringBuilder().append("Ingesting with: ")
+                    .append("dataFormat - ").append(dataFormat).append("|")
+                    .append("ingestionMappingName - ").append(mappingName).append("|")
+                    .append("databaseName - ").append(databaseName).append("|")
+                    .append("tableName - ").append(tableName).append("|")
+                    .append("pollOnIngestionStatus - ").append(pollOnIngestionStatus).append("|")
+                    .append("ingestionStatusPollingTimeout - ").append(ingestionStatusPollingTimeout).append("|")
+                    .append("ingestionStatusPollingInterval - ").append(ingestionStatusPollingInterval).append("|")
+                    .append("routePartiallySuccessfulIngestion - ").append(routePartiallySuccessfulIngestion);
+            getLogger().debug(ingestLogString.toString());
```
This StringBuilder approach is not efficient because it requires a processing for concatenation just to create a debug log. The StringBuilder should be removed and replaced with a parameterized log statement.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+                    new KustoIngestionRequest(streamingEnabled, pollOnIngestionStatus, inputStream, databaseName,
+                            tableName, dataFormat, mappingName, ignoreFirstRecord, ingestionStatusPollingTimeout, ingestionStatusPollingInterval)
+            );
+            getLogger().info("Ingest Status Polling Enabled {} and Ingest Status Polling Enabled {} ", pollOnIngestionStatus, result.toString());
```
The `result.toString()` is not necessary. This seems more appropriate as a debug log.
```suggestion
            getLogger().debug("Ingest Status Polling Enabled [{}] and Ingest Status Polling Enabled {}", pollOnIngestionStatus, result);
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            } else if (result == KustoIngestionResult.PARTIALLY_SUCCEEDED) {
+                getLogger().warn("Ingest Partially succeeded - {}", result.getStatus());
+                flowFile = session.putAttribute(flowFile, "ingestion_status", "partial_success");
+                if (StringUtils.equalsIgnoreCase(routePartiallySuccessfulIngestion, "Failure")) {
```
This should be compared using the failure relationship name instead of repeating the `Failure` string for the comparison.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+
+        } catch (IOException | URISyntaxException e) {
+            getLogger().error("Azure Data Explorer Ingest processing failed", e);
+            isError = true;
```
The `isError` variable is not very clear. Instead, recommend naming it `transferRelationship` and setting the appropriate relationship value.

tanmaya-panda1 replied:
removed isError variable

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+            throw new ProcessException(String.format("Error occurred while checking if streaming policy is enabled for the table for entity %s in database %s",entityName,database));
+        }
+        List<List<Object>> queryResult = kustoQueryResponse.getQueryResult();;
+        if ( queryResult.get(0) !=null && queryResult.get(0).get(2)!=null && StringUtils.isNotEmpty(queryResult.get(0).get(2).toString())) {
```
This conditional is difficult to follow based on the construction and formatting. The `List<List<Object>>` structure is also very generic, which makes this difficult to understand. If this is a value that should always be checked, recommend evaluating the value in the service and providing a separate named property.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,309 @@
+    protected boolean checkIfStreamingPolicyIsEnabledInADX(String entityName, String database) {
+        KustoQueryResponse kustoQueryResponse = service.executeQuery(database, String.format(STREAMING_POLICY_SHOW_COMMAND, PutAzureDataExplorer.DATABASE, entityName));
+        if (kustoQueryResponse.isError()) {
+            throw new ProcessException(String.format("Error occurred while checking if streaming policy is enabled for the table for entity %s in database %s",entityName,database));
```
```suggestion
            throw new ProcessException(String.format("Error occurred while checking if streaming policy is enabled for the table for entity %s in database %s", entityName, database));
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,362 @@
+            APPLICATION_TENANT_ID,
+            CLUSTER_URI));
+
+    public static final Pair<String, String> NIFI_SINK = Pair.of("processor", "nifi-sink");
```
Instead of using the word `nifi-sink`, it seems better to use the component class simple name.
```suggestion
    public static final Pair<String, String> NIFI_SINK = Pair.of("processor", StandardKustoIngestService.class.getSimpleName());
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,362 @@
+
+    /**
+     * @param context the configuration context
+     * @throws InitializationException if unable to create a database connection
```
This `throws` line is not correct. It is probably better to remove the Javadoc comment if it is not completely written.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,293 @@
+                List<IngestionStatus> statuses;
+                long startTime = System.currentTimeMillis();
+                long timeoutMillis = 600 * 1800; // 10 Minutes timeout
+                while (true) {
```
Please review this recommendation to adjust the loop approach.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,362 @@
+            if (kustoIngestionRequest.pollOnIngestionStatus()) {
+                List<IngestionStatus> statuses;
+                long startTime = System.currentTimeMillis();
+                long timeoutMillis = Long.parseLong(kustoIngestionRequest.getIngestionStatusPollingTimeout()) * 1000L;
```
Instead of parsing the value in this method, the KustoIngestRequest should use a `java.time.Duration` and the value should be parsed in the Processor.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,362 @@
+                    }
+
+                    // Sleep for 5 seconds before checking again
+                    Thread.sleep(Integer.parseInt(kustoIngestionRequest.getIngestionStatusPollingInterval()) * 1000L);
```
See the related note on parsing this value in the calling method and using `java.time.Duration`.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-services-api/src/main/java/org/apache/nifi/services/azure/data/explorer/KustoIngestionRequest.java:
```diff
@@ -0,0 +1,100 @@
+
+    private String ignoreFirstRecord;
+
+    private String ingestionStatusPollingTimeout;
+
+    private String ingestionStatusPollingInterval;
```
As noted elsewhere, these properties should be changed to `java.time.Duration`.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,362 @@
+                kustoResultSetTable.next();
+                kustoQueryResponse = new KustoQueryResponse(kustoResultSetTable.getData());
+            } else {
+                kustoQueryResponse = new KustoQueryResponse(new ArrayList<>());
```
```suggestion
                kustoQueryResponse = new KustoQueryResponse(Collections.emptyList());
```

exceptionfactory reviewed (changes requested):
Thanks for your patience on the review @tanmaya-panda1. The last round of updates were helpful in clearing up some of the implementation.

I noted several detailed adjustments, but there is one high-level concern that should be revisited. Returning the generic `List<List<Object>>` of query results is difficult to follow in the context of the ingest status query, and it could introduces the data parsing that we intentionally avoided for the previous query processor. Instead of reusing the existing `KustoQueryResponse executeQuery()` method on the `KustoQueryService` interface, this seems like an opportunity to return a well-formed Ingest Status response, which would allow the `StandardKustoIngestService` implementation to handle the internals of parsing the information before returning it to the Processor.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,291 @@
+    private static final List<PropertyDescriptor> descriptors;
+    private static final Set<Relationship> relationships;
+    private transient KustoIngestService service;
+    private boolean streamingEnabled;
+    private boolean pollOnIngestionStatus;
```
These values should not be member variables.
```suggestion
```

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-services-api/src/main/java/org/apache/nifi/services/azure/data/explorer/KustoQueryResponse.java:
```diff
@@ -27,16 +28,27 @@ public class KustoQueryResponse {
 
     private final String errorMessage;
 
+    private final List<List<Object>> queryResult;
```
Now that the interface has been redefined, can this list of results be removed? The values should be passed in the service implementation methods, which would avoid adding this member variable to the public response object.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/data/explorer/StandardKustoIngestService.java:
```diff
@@ -0,0 +1,404 @@
+                ingestionResult = queuedIngestClient.ingestFromStream(info, ingestionProperties);
+            }
+            if (kustoIngestionRequest.pollOnIngestionStatus()) {
+                List<IngestionStatus> statuses = initializeKustoIngestionStatusAsPending();
+                long startTime = System.currentTimeMillis();
+                long timeoutMillis = kustoIngestionRequest.getIngestionStatusPollingTimeout().toMillis();
+                // Calculate the end time based on the timeout duration
+                long endTime = startTime + timeoutMillis;
+                while (System.currentTimeMillis() < endTime) {
+                    // Get the status of the ingestion operation
+                    List<IngestionStatus> statuses1 = ingestionResult.getIngestionStatusCollection();
+                    if (statuses1.get(0).status == OperationStatus.Succeeded
+                            || statuses1.get(0).status == OperationStatus.Failed
+                            || statuses1.get(0).status == OperationStatus.PartiallySucceeded) {
+                        statuses = statuses1;
+                        break;
+                    }
+                    // Sleep for before checking again
+                    Thread.sleep(kustoIngestionRequest.getIngestionStatusPollingInterval().toMillis());
+                }
+                // Check if the timeout has been exceeded
+                if (System.currentTimeMillis() - startTime >= timeoutMillis) {
+                    throw new ProcessException(String.format("Timeout of %s exceeded while waiting for ingestion status", kustoIngestionRequest.getIngestionStatusPollingTimeout()));
+                }
+                return KustoIngestionResult.fromString(statuses.get(0).status.toString());
```
Recommend moving this block into a separate method named something like `pollIngestionStatus()`.

exceptionfactory reviewed (changes requested):
Thanks for the updates @tanmaya-panda1. I noted a few more naming and interface definition recommendations, but this looks closer to completion. It looks like it should be ready in one or two more iterations.

tanmaya-panda1 commented:
@exceptionfactory : Can you please take a look at this PR. This is long pending item, we wanted to close this before the end of this year. We would appreciate your review on this item whenever you have some time.

exceptionfactory commented:
> @exceptionfactory : Hope you are doing good. Can you please take a look at this PR. This is long pending item, we wanted to close this before the end of this year. We would appreciate your review on this item whenever you have some time.

Thanks for the follow-up @tanmaya-panda1. I am in the process of reviewing several pull requests, and I plan to return to this one soon.

exceptionfactory commented on nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/data/explorer/PutAzureDataExplorer.java:
```diff
@@ -0,0 +1,282 @@
+            .identifiesControllerService(KustoIngestService.class)
+            .build();
+    public static final PropertyDescriptor DATA_FORMAT = new PropertyDescriptor.Builder()
+            .name("Data Format")
```
What do you think about supporting Expression Language for the Data Format property? This would allow the property value to be resolved based on a FlowFile attribute, so that one Processor could handle different types of data. That would require call `evaluateAttributeExpressions(flowFile)` in the `onTrigger()` method.

tanmaya-panda1 replied:
yeah makes sense.

exceptionfactory reviewed (commented):
Thanks again for your patience and work @tanmaya-panda1. This looks close to completion, I plan to go through a few runtime tests soon before wrapping it up. I noted one more feature level question on the `Data Format` property, which could make the Processor configuration much more flexible when dealing with different types of data. That is not absolutely required.

exceptionfactory reviewed (commented):
@tanmaya-panda1 Thanks for your work and patience on this pull request. I went through the latest version. As the main branch has changed a good deal since the initial pull request, I rebased the pull request to incorporate the current baseline. In the course of reviewing the final version, I made a few adjustments to the Controller Service interface definition to provide better encapsulation of streaming policy and table readability operations.

With these changes, this should be ready to merge very soon.

exceptionfactory commented:
Rebased to resolve merge conflicts.

exceptionfactory reviewed (approved):
Thanks again for your work and patience on this pull request @tanmaya-panda1. The latest version checks out in runtime testing, so I will proceed with merging. +1
